### PR TITLE
CMake: Fix aarch64 vectorization support.

### PR DESCRIPTION
**Pull request by teknoman117**
- Fix detection of aarch64 for OPUS_CPU_ARM and adding sources from
  celt_sources_arm (previously would miss armcpu.c and arm_celt_map.c)

- Rename OPUS_MAY_SUPPORT_NEON to OPUS_MAY_HAVE_NEON as this name is
  used everywhere else in the CMake build system. Without this,
  runtime capability detection is broken on aarch64.

### CODE REVIEW

**Comment by teknoman117** [comment]
For testing android, you can do the following:
ARMv7 (older android platform)
`
cmake \
    -DCMAKE_TOOLCHAIN_FILE=/path/to/ndk-bundle/build/cmake/android.toolchain.cmake \
    -DANDROID_ABI=armeabi-v7a \
    -DANDROID_ARM_NEON=ON \
    -DANDROID_PLATFORM=android-21 \
    -GNinja \
    -DCMAKE_BUILD_TYPE=Release \
    -DBUILD_SHARED_LIBS=ON ..
`

ARMv7 (current android platform) - android-23 and higher automatically enables support for NEON on ARMv7
`
cmake \
    -DCMAKE_TOOLCHAIN_FILE=/path/to/ndk-bundle/build/cmake/android.toolchain.cmake \
    -DANDROID_ABI=armeabi-v7a \
    -DANDROID_PLATFORM=android-29 \
    -GNinja \
    -DCMAKE_BUILD_TYPE=Release \
    -DBUILD_SHARED_LIBS=ON ..
`

ARMv8 (current android platform)
`
cmake \
    -DCMAKE_TOOLCHAIN_FILE=/path/to/ndk-bundle/build/cmake/android.toolchain.cmake \
    -DANDROID_ABI=arm64-v8a \
    -DANDROID_PLATFORM=android-29 \
    -GNinja \
    -DCMAKE_BUILD_TYPE=Release \
    -DBUILD_SHARED_LIBS=ON ..
`

File: opus_config.cmake [hunk 1]
```diff
@@ -34,7 +34,7 @@ elseif(OPUS_CPU_ARM)
   opus_detect_neon(COMPILER_SUPPORT_NEON)
   if(COMPILER_SUPPORT_NEON)
     option(OPUS_USE_NEON "Option to turn off SSE" ON)
```
Comment (xnorpx) [inline opus_config.cmake]: can you fix the typo in the comment on line 36 SSE->NEON

File: opus_functions.cmake [hunk 2]
```diff
@@ -202,6 +202,12 @@ function(opus_supports_cpu_detection RUNTIME_CPU_CAPABILITY_DETECTION)
   endif()
   if(HAVE_INTRIN_H OR HAVE_CPUID_H)
     set(RUNTIME_CPU_CAPABILITY_DETECTION 1 PARENT_SCOPE)
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(armv7-a|aarch64)")
+    # ARM cpu detection is implemented for Windows and anything
+    # using a Linux kernel (such as Android).
+    if (CMAKE_SYSTEM_NAME MATCHES "(Windows|Linux|Android)")
```
Comment (xnorpx) [inline opus_functions.cmake]: lets add iOS as well

Comment (xnorpx) [reply opus_functions.cmake]: https://cmake.org/cmake/help/latest/variable/IOS.html

**Comment by xnorpx** [comment]
Added some comments, I would like to add support to build Android flavors in the github actions as well to test it. I am quite busy this week but will get back to this early next week. Thanks for the help!

**Comment by xnorpx** [comment]
if you want you can create a pull request to create the build config for android in the github actions ass well. Otherwise I do it next week to verify these changes.

**Comment by xnorpx** [comment]
I added iOS, it should be straight forward to add Android once one find the path to the NDK.

**Comment by teknoman117** [comment]
> if you want you can create a pull request to create the build config for android in the github actions as well. Otherwise I do it next week to verify these changes.

I've not done this before, what repo drives the github actions?

> I added iOS, it should be straight forward to add Android once one find the path to the NDK.

https://developer.android.com/ndk/downloads

**Comment by xnorpx** [comment]
It's just part of this repo, here is the yaml file: https://github.com/xnorpx/opus/blob/master/.github/workflows/build.yml

**Comment by xnorpx** [comment]
Here is the result: https://github.com/xnorpx/opus/actions

**Comment by xnorpx** [comment]
I just merged android build, please refresh your branch and see how it works

File: opus_functions.cmake (hunk 2 above)
Comment (teknoman117) [reply opus_functions.cmake:208]: The reason I didn't add it here was that it doesn't appear that runtime capability detection is implemented for iOS.

**Comment by teknoman117** [comment]
I fixed the typo in the NEON option. I didn't add iOS to the list of ARM platforms which allow runtime capability detection because I didn't see an implementation for iOS. That being said, I think you could reasonable set OPUS_PRESUME_NEON on iOS because all iPhones with the exception of the first two ("iPhone" and iPhone 3G) supported NEON.

**Comment by xnorpx** [comment]
I see, good point we should just set OPUS_PRESUME_NEON for apple arm devices

**Comment by xnorpx** [comment]
Please squash into one commit then we can merge it

**Comment by teknoman117** [comment]
made the iOS change, squashed, and force-pushed.

**Comment by teknoman117** [comment]
out of curiosity, one what general cycle do your CMake changes get pulled into the main opus repo?

**Comment by teknoman117** [comment]
I also changed the instances of "armv7-a" to "arm" because systems such as the RPi 3 report "CMAKE_SYSTEM_PROCESSOR" as armv7l

**Comment by xnorpx** [comment]
"out of curiosity, on what general cycle do your CMake changes get pulled into the main opus repo?"

So far as time permits, but no with CI build setup and we get this major bugs fix it should hopefully be faster. but essentially. My time then I send patches to jean-marc who merges them into Xiph repo. After that it gets merged back to master github mirror. But hopefully I can be faster now.

**Comment by xnorpx** [comment]
"I also changed the instances of "armv7-a" to "arm" because systems such as the RPi 3 report "CMAKE_SYSTEM_PROCESSOR" as "armv7l". It also appears that opus supports older ARM revisions than ARMv7 (e.g. https://github.com/xnorpx/opus/blob/master/celt/arm/kiss_fft_armv4.h)"

Ok sounds good. will take a new look tonight and do the merge.

**Pull request merged**